lib.meta.getLicenseFromSpdxIdOr: init

ShamrockLee opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

This PR introduces `lib.meta.getLicenseFromSpdxIdOr`, a variant function of `lib.meta.getLicenseFromSpdxId` that allows the developers to specify the fallback value explicitly when a license with a matching SPDX ID is not found in `lib.licenses`.

Given that the typical use case of this function is to parse configuration files such as `package.json` and `pyproject.toml` during automatic package generation, the warning introduced by `lib.getLicenseFromSpdxId` will block the Release Tests and OfBorg evaluation for a programmatically generated package set.

`lib.getLicenseFromSpdxIdOr` enables developers to control the fallback behavior, leading to a smoother developer and user experience.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

ShamrockLee commented:
@infinisil Do you have time taking a look at this?

ShamrockLee commented:
Rebase and solve the merge conflict.

ShamrockLee commented:
Bump!

infinisil commented on lib/meta.nix:
```diff
@@ -139,18 +140,34 @@ rec {
+     lib.getLicenseFromSpdxId' "MY LICENSE" (builtins.abort "No SPDX ID matches MY LICENSE")
+     => error: evaluation aborted with the following error message: 'No SPDX ID matches MY LICENSE'
+  */
+  getLicenseFromSpdxId' =
```
```suggestion
  getLicenseFromSpdxIdOr =
```

This is then similar to `spdxLicenses.${ ... } or default`

infinisil replied:
I'd say let's just introduce the -`Or` variant for now. One step at a time :)

ShamrockLee replied:
> I'd say let's just introduce the -Or variant for now. One step at a time :)

Addressed.

infinisil reviewed (changes requested):
I have to agree that functions shouldn't throw a warning if you can't know whether the warning would be thrown without calling it.

If I could roll back time, I would've wanted `getLicenseFromSpdxId` to throw an error in case there's no match. Implicit defaults are not a great thing. I'd be interested in deprecating this function in the future, instead preferring `getLicenseFromSpdxIdOr`.

ShamrockLee committed:
lib.getLicenseFromSpdxIdOr: init

Add lib.meta.getLicenseFromSpdxIdOr as a variant of
lib.meta.getLicenseFromSpdxId that explicitly state the default
(fallback) value if there's no license matching the given SPDX ID.

infinisil commented on lib/meta.nix:
```diff
@@ -147,15 +146,36 @@ rec {
+     Type:
+       getLicenseFromSpdxIdOr :: str -> Any -> Any
+
+     Example:
```
Could you add some tests in `lib/tests/misc.nix`? These examples would work well :)

ShamrockLee replied:
Tests added for `getLicenseFromSpdxIdOr`

infinisil reviewed (commented):
Other than missing tests, this is looking good!

ShamrockLee committed:
lib.getLicenseFromSpdxId: improve documentation

ShamrockLee commented:
I added tests for `lib.meta.getLicenceFromSpdxIdOr` and modernized its documentation.

ShamrockLee commented:
@infinisil, the test is added. Could you help review it again?

infinisil reviewed (approved):
Ah yes, sorry for the delay, great work!
